### PR TITLE
fix: address domain separation regarding challenge generation for mac origin (see issue #4333)

### PR DESCRIPTION
**Pull request by jorgeantonio21**
Description
--- 
Add use of hashing API to create challenge for mac origin.

Motivation and Context
--- 
Tackle [issue](https://github.com/tari-project/tari/issues/4333).

How Has This Been Tested?
--- 
Existing unit tests.

### CODE REVIEW

File: comms/dht/src/inbound/decryption.rs [hunk 1]
```diff
@@ -248,7 +248,7 @@ where S: Service<DecryptedDhtMessage, Response = (), Error = PipelineError>
             Ok(origin_mac) => {
                 // If this fails, discard the message because we decrypted and deserialized the message with our shared
                 // ECDH secret but the message could not be authenticated
-                let msg_challenge = crypt::create_message_challenge(&message.dht_header, &message.body);
+                let msg_challenge = crypt::create_message_domain_separated_hash(&message.dht_header, &message.body);
```
Comment (AaronFeickert) [inline comms/dht/src/inbound/decryption.rs]: Recommend renaming to be consistent with the rest of the PR.

Comment (AaronFeickert) [reply comms/dht/src/inbound/decryption.rs]: Something like `binding_message_representation` might be nicely expressive if used consistently.

Comment (jorgeantonio21) [reply comms/dht/src/inbound/decryption.rs]: I renamed these variables, as you suggested, as well as the terminology `OriginMac`/`origin_mac`

File: comms/dht/src/crypt.rs [hunk 2]
```diff
@@ -43,6 +43,8 @@ use crate::{
     version::DhtProtocolVersion,
 };
 
+const DOMAIN_SEPARATION_CHALLENGE_LABEL: &str = "com.tari.comms.dht.crypt.message_parts";
```
Comment (hansieodendaal) [inline comms/dht/src/crypt.rs:46]:
```suggestion
```

Please use `comms_dht_hash_domain()`

File: comms/dht/src/crypt.rs [hunk 3]
```diff
@@ -127,10 +125,25 @@ pub fn create_message_challenge_parts(
...
         })
         .unwrap_or_default();
-    mac_challenge.update(&e_pk);
 
-    mac_challenge.update(&body);
-    mac_challenge.finalize_fixed().into()
+    // we digest the given data into a domain independent hash function to produce a signature
+    // use of the hashing API for domain separation and deal with variable length input
+    let domain_separated_hash =
+        DomainSeparatedHasher::<Challenge, GenericHashDomain>::new(DOMAIN_SEPARATION_CHALLENGE_LABEL)
```
Comment (hansieodendaal) [inline comms/dht/src/crypt.rs:132]:
```suggestion
        comms_dht_hash_domain().mac_hasher::<Challenge>()
```
Note that `mac_hasher` can also be used, which implements the `LengthExtensionAttackResistant` trait.

File: comms/dht/src/crypt.rs [hunk 4]
```diff
@@ -127,10 +125,25 @@ pub fn create_message_challenge_parts(
...
+            .chain(&expires)
+            .chain(&e_pk)
+            .chain(&body)
+            .finalize()
+            .into_vec();
+
+    let mut output = [0u8; 32];
+    // the use of Challenge bind to Blake256, should always produce a 32-byte length output data
+    output.copy_from_slice(domain_separated_hash.as_slice());
+    output
```
Comment (hansieodendaal) [inline comms/dht/src/crypt.rs:146]:
```suggestion
            .finalize();
        
    domain_separated_hash.hash_to_bytes()
        .expect("Fixed output of '0 < size <= 32' bytes with 32 byte hasher cannot fail.")
```
This way the runtime panic is not hidden in `copy_from_slice`. As an alternative, the function could return a result and the runtime error could be handled.

File: comms/dht/src/crypt.rs [hunk 5]
```diff
@@ -102,7 +104,7 @@ pub fn create_message_challenge(header: &DhtMessageHeader, body: &[u8]) -> [u8;
 }
 
 /// Generates a 32-byte hashed challenge that commits to all message parts
-pub fn create_message_challenge_parts(
+pub fn create_message_domain_separated_hash_parts(
```
Comment (hansieodendaal) [inline comms/dht/src/crypt.rs:107]:
```suggestion
pub fn create_message_hash(
```
The previous name conveys the intent well enough, but the addition of `_parts` is confusing, so `create_message_challenge` could also work.

File: comms/dht/src/dedup/mod.rs [hunk 6]
```diff
@@ -44,11 +44,11 @@ use crate::{
...
 
 pub fn hash_inbound_message(msg: &DhtInboundMessage) -> [u8; 32] {
-    create_message_hash(&msg.dht_header.origin_mac, &msg.body)
+    create_message_hash(&msg.dht_header.message_signature, &msg.body)
 }
 
-pub fn create_message_hash(origin_mac: &[u8], body: &[u8]) -> [u8; 32] {
-    Challenge::new().chain(origin_mac).chain(&body).finalize().into()
+pub fn create_message_hash(message_signature: &[u8], body: &[u8]) -> [u8; 32] {
+    Challenge::new().chain(message_signature).chain(&body).finalize().into()
```
Comment (hansieodendaal) [inline comms/dht/src/dedup/mod.rs:51]:
```suggestion
        comms_dht_hash_domain()
        .mac_hasher::<Challenge>()
        .chain(message_signature)
        .chain(&body)
        .finalize()
        .hash_to_bytes()
        .expect("Fixed output of '0 < size <= 32' bytes with 32 byte hasher cannot fail.")
```

File: comms/dht/src/message_signature.rs [hunk 7]
```diff
@@ -29,12 +29,12 @@ use tari_crypto::keys::PublicKey;
...
 
 #[derive(Debug, Clone)]
-pub struct OriginMac {
+pub struct MessageSignature {
     signer_public_key: CommsPublicKey,
     signature: Signature,
 }
 
-fn construct_origin_mac_hash(
+fn construct_message_signature_hash(
```
Comment (hansieodendaal) [inline comms/dht/src/message_signature.rs:37]: This function should also implement `comms_dht_hash_domain()` as 

``` rust
pub fn create_message_hash(message_signature: &[u8], body: &[u8]) -> [u8; 32] {
    comms_dht_hash_domain()
        .hasher::<Challenge>()
        .chain(message_signature)
        .chain(&body)
        .finalize()
        .hash_to_bytes()
        .expect("Fixed output of '0 < size <= 32' bytes with as 32 byte hasher cannot fail.")
}
```

**Review by hansieodendaal — CHANGES_REQUESTED** [review]
Looking good, please see suggestions below.

**Comment by jorgeantonio21** [comment]
This PR has been addressed by [this](https://github.com/tari-project/tari/pull/4336).

**Pull request closed without merging**